File: 01-entrevistas-system-design/01-rate-limiter/node/token-bucket.ts
/**
 * Token Bucket Rate Limiter using Node.js (Single Threaded)
 * Reference: https://en.wikipedia.org/wiki/Token_bucket
 *
 */

export class TokenBucket {
  private capacity: number; // maximum number of tokens in the bucket
  private tokens: number;
  private refillRate: number; // number of tokens to add per second
  private lastRefillTimestamp: number;

  constructor(capacity: number, refillRate: number) {
    this.capacity = capacity;
    this.tokens = capacity; // start full
    this.refillRate = refillRate;
    this.lastRefillTimestamp = Date.now();
  }

  getTokens(): number {
    return this.tokens;
  }

  /**
   * This method is called for each incoming request to check if it can be allowed.
   *
   * @returns
   */
  allowRequest(): boolean {
    this.refillTokens();

    if (this.tokens >= 1) {
      this.tokens -= 1;
      return true; // request allowed
    }

    return false; // request denied: 429 Too Many Requests
  }

  /**
   * Refill tokens based on the elapsed time since the last refill.
   */
  private refillTokens() {
    const now = Date.now();
    const elapsedTime = (now - this.lastRefillTimestamp) / 1000;

    // Calculate how many tokens to add based on elapsed time and refill rate
    const tokensToAdd = elapsedTime * this.refillRate;

    // Refill tokens, but never exceed capacity
    this.tokens = Math.min(this.capacity, this.tokens + tokensToAdd);
    this.lastRefillTimestamp = now;
  }
}

File: 01-entrevistas-system-design/01-rate-limiter/go/token_bucket.go
package main

import (
	"math"
	"sync"
	"time"
)

// TokenBucket Rate Limiter
// Reference: https://en.wikipedia.org/wiki/Token_bucket
type TokenBucket struct {
	capacity            float64                   // Maximum number of tokens in the bucket
	tokens              float64                   // Current number of tokens
	refillRate          float64                   // Number of tokens to add per second
	lastRefillTimestamp time.Time                 // Timestamp of the last refill
	mu                  sync.Mutex                // Mutex for thread-safe access
}

// NewTokenBucket creates a new TokenBucket with the given capacity and refill rate.
func NewTokenBucket(capacity float64, refillRate float64) *TokenBucket {
	return &TokenBucket{
		capacity:            capacity,
		tokens:              capacity, // start full
		refillRate:          refillRate,
		lastRefillTimestamp: time.Now(),
	}
}

// GetTokens returns the current number of tokens.
func (tb *TokenBucket) GetTokens() float64 {
	tb.mu.Lock()
	defer tb.mu.Unlock()
	return tb.tokens
}

// AllowRequest checks if a request can be allowed.
// It returns true if allowed, false otherwise.
func (tb *TokenBucket) AllowRequest() bool {
	tb.mu.Lock()
	defer tb.mu.Unlock()

	tb.refillTokens()

	if tb.tokens >= 1 {
		tb.tokens -= 1
		return true // request allowed
	}

	return false // request denied: 429 Too Many Requests
}

// refillTokens updates the token count based on elapsed time.
func (tb *TokenBucket) refillTokens() {
	now := time.Now()
	elapsedTime := now.Sub(tb.lastRefillTimestamp).Seconds()

	// Calculate how many tokens to add based on elapsed time and refill rate
	tokensToAdd := elapsedTime * tb.refillRate

	// Refill tokens, but never exceed capacity
	tb.tokens = math.Min(tb.capacity, tb.tokens+tokensToAdd)
	tb.lastRefillTimestamp = now
}
